Feature fielddb2.50 integration

cesine opened the pull request:
Step 1:

* loading FieldDB via requirejs
* changing the brand depending on the server the user selected
* using the Fielddb auth login methods
* implementing the confirm, bug and prompt UI hooks

cesine commented on app/scripts/models/application-settings.coffee:
```diff
@@ -22,6 +23,7 @@ define [
 
     initialize: ->
+      @set('fielddbApplication' , FieldDB.FieldDBObject.application)
       @listenTo Backbone, 'authenticate:login', @authenticate
```
@jrwdunham what is the syntax to do this line so we can access the fielddbapplication later via
globals.applicationSettings.fielddbApplication 
or 
globals.applicationSettings.get('fielddbApplication')

jrwdunham replied:
The second one.

jrwdunham replied:
In app/scripts/collections/forms.coffee you removed the `pouchname` from `"#{url}/#{pouchname}/_design/pages/_view/datums_chronological"`. Why did you do that? Is it because the `url` returned by the commonjs fielddb includes the pouchname of the active corpus?

Also, you're accessing the `fiedldbApplication` as an attribute and not a BB model attribute (i.e., via `.get`), which is what you need to do.

cesine commented:
@jrwdunham  please review and test and merge this one first so that i can get you feedback before moving on to the corpus.coffee

cesine commented on app/scripts/collections/forms.coffee:
```diff
@@ -205,9 +205,8 @@ define [
-      "#{url}/#{pouchname}/_design/pages/_view/datums_chronological"
+      url = globals.applicationSettings.fielddbApplication.corpus.url
+      "#{url}/_design/pages/_view/datums_chronological"
 
```
@jrwdunham  yes, a fielddb corpus knows its (current) url (which includes its pouchname), it should be able to take any url, eg if the couchdb is running in a virtual directory, or if its not a couchdb at all but responds to get and post and put requests

```javascript
 url: {
    get: function() {
      if (this.connection && this.connection.corpusUrl) {
        return this.connection.corpusUrl;
      } else {
        if (this.dbname) {
          return this.BASE_DB_URL + "/" + this.dbname;
        }
      }
      this.warn("Using an unlikely url, as if this app was running in a website where the databse is.");
    },

```

cesine commented:
@jrwdunham  i removed more cruft, its resuming auth sessions on my computer and other events are working (ie, you login once, and refresh and it knows who you are) it doesnt show your list of corpora yet because i need want to get the auth approved before i move on to the corpora select process.

cesine commented:
@jrwdunham please click the restart button on the build. there was a dependacny missing in fielddb when the last commit was sumbitted, its there now.

cesine commented:
@jrwdunham if you're still awake can you give some feedback on this pull request... all other apps have been updated to use the commonjs library now, if you don't start using it soon you will be receiving data you have to do extra work to handle. it also reduces the number of server requests.

jrwdunham commented on app/scripts/collections/forms.coffee:
```diff
@@ -205,9 +205,8 @@ define [
-      url = globals.applicationSettings.get 'baseDBURL'
-      pouchname = globals.applicationSettings.get 'activeFieldDBCorpus'
-      "#{url}/#{pouchname}/_design/pages/_view/datums_chronological"
+      url = globals.applicationSettings.get('fieldDBApplication').corpus.url
```
The `url` you get here: it has the pouchname in it? I.e., it's <url>/<pouchname>/ ? Just checking ...

cesine replied:
yes

jrwdunham commented on app/scripts/models/application-settings.coffee:
```diff
@@ -399,40 +305,36 @@ define [
-        type: 'FieldDB'
-        url: 'https://localhost:3183'
-        serverCode: 'localhost'
-        corpusServerURL: null
```
How are you doing this without persisting `corpusServerURL`? All of the FieldDB-oriented authentication logic that you've deleted in this commit was designed to allow a FieldDB web service to be specified by only providing an "auth server url". Under the new authentication logic you're implementing, will the user now need to specify both an "auth server url" and a "corpus server url" for FieldDB connections? If so, then [ a)I don't like that and] b) we need to change the server interface accordingly (if you haven't done so already...)

cesine replied:
This code is equivalent to what you had.

 Basically, we moved the code that knows the usual server combinations, and knows how to guess server combinations into the commonjs so you can use it. I also incorporated some of your feedback about offline replication urls and logging the user into their corpus server when they login to the auth service.
https://github.com/OpenSourceFieldlinguistics/FieldDB/blob/master/api/corpus/Connection.js#L560-570

This shows how it works: https://github.com/OpenSourceFieldlinguistics/FieldDB/blob/master/tests/corpus/DatabaseTest.js#L438-468

Here is an example of how its used:
https://github.com/OpenSourceFieldlinguistics/AuthenticationWebService/blob/master/lib/nodeconfig_local.js#L19-37

jrwdunham commented on app/scripts/views/app.coffee:
```diff
@@ -569,10 +564,22 @@ define [
       # TODO @jrwdunham @cesine: figure out how i18n/localization works in
       # FieldDB and begin implementing something similar in Dative.
-      console.log ["TODO show some visual thing here using the app view using",
+      console.log ["Pretending the user said Yes. This is dangerous. TODO show some visual confirm dialog here using the app view using",
```
This weird string syntax was before I understood how CoffeeScript multiline strings work. You can just add linebreaks and proper indentation in the middle of the string and it will all work out, i.e., remove the joined array.

cesine replied:
k  i also updated the other string above this one

jrwdunham commented on app/scripts/views/app.coffee:
```diff
@@ -569,10 +564,22 @@ define [
-      console.log ["TODO show some visual thing here using the app view using",
+      console.log ["Pretending the user said Yes. This is dangerous. TODO show some visual confirm dialog here using the app view using",
         "something like http://jqueryui.com/dialog/#modal-confirmation" ,
         message].join ' '
```
Dative has a modal confirm widget that can be used for this. It's the `AlertDialogView`, I think. See an example of its usage in the `FormView` when the delete/destroy button is clicked.

cesine replied:
awesome! i found the code and implemented it. the only problem i see is that your alert dialog doesnt handle cancel events? i will try to add it since the fielddb confirms use it.

jrwdunham commented on app/scripts/views/app.coffee:
```diff
@@ -569,10 +564,22 @@ define [
 
+      # NOTE cesine: overiding confirm MUST return a promise, otherwise its broken.
+      deferred = FieldDB.Q.defer()
+      self = this
```
In general, you shouldn't need to write `self = this` in CoffeeScript. Anyways, you're not even using `self` here ...

cesine commented:
@jrwdunham  i did the changes and added some notes for you. I also implemented the bug, confirm and prompt dialogs so you have some extra code to look at in terms of coffeescriptification.

cesine commented:
please take a look at this if you have time today :) i'm not jet lagged yet so i can work on this stuff if you want

cesine commented:
need to make the connection return the authUrl as
![screen shot 2015-05-12 at 1 05 02 pm](https://cloud.githubusercontent.com/assets/196199/7593433/1d0d6a0c-f8a8-11e4-9bad-f95c6c318a44.png)
 url so you can edit it in the app settings

cesine commented:
@jrwdunham even setting server1 to an object in Dative i cant get the `url` to be set by the time it gets to the serverscollection.

![screen shot 2015-05-14 at 5 23 14 pm](https://cloud.githubusercontent.com/assets/196199/7642389/949b5d7e-fa5e-11e4-9c8a-40bc607400e9.png)

( i gave up on using the fielddb connections because i found that  the Server class wont accept any extra attributes anyway)

![screen shot 2015-05-14 at 5 27 47 pm](https://cloud.githubusercontent.com/assets/196199/7642402/a1e884de-fa5e-11e4-823f-67b3c2df7646.png)

jrwdunham commented:
@cesine logging out from FieldDB is not working. Under `logoutFieldDB` at line 272 of models/application-settings.coffee, `@get('fieldDBApplication').authentication.logout` is undefined.

cesine commented:
can you paste the version of fielddb here? maybe you didnt run the grunt update tasks that i made for you. just incase you can also paste the contents of @get('fieldDBApplication') here to see if its actually defined too

cesine commented:
@jrwdunham  i think i might have found the problem. If we call fetch() in the initialize, then the @get('fieldDBApplication') wont be empty:
![screen shot 2015-06-16 at 7 06 36 pm](https://cloud.githubusercontent.com/assets/196199/8196582/2b7a9c6c-145b-11e5-9c4c-3c5e500fa518.png)

Does this seem likely and/or feasible as a solution? or is there a fetch() somewhere prior/after to the initialize that should have been called?
